README.md
Parses Papertrail logs to show context and historical reflection of Python assertions.

At a high level, here's the steps we perform:
- Papertrail generates archives of all its logs once per hour
- S3 triggers a Lambda function that calls our API (passing along the new log file's filename)
- we download the log file and parse it for Python tracebacks. we save the tracebacks to ElasticSearch
- we provide a Jinja templated site to show these tracebacks to our users

[![Build Status](https://travis-ci.org/topher200/assertion-context.svg?branch=master)](https://travis-ci.org/topher200/assertion-context)

# Parsing Papertrail logs

We take the archived Papertrail logs in .tsv.gz format and unzip them. We search
them for any errors. When an error is found, we go backwards in the logs to find
the root cause (context) of the assertion, which includes the traceback and the
log lines before the traceback. We do this by determining the instance ID of the
machine that had the assertion and searching the logs for the lines on that
machine previous to the offending line.

# Getting it set up
## Server installation instructions
We run everything in kubernetes. Please set up kubectl locally and connect to a 
kubernetes cluster. Then, on your development machine...
 - install kubernetes-helm
 - clone this repo
 - fill out .env with config variables. requires these fields:
```
DEBUG_LOGGING = False  # enables logging to file, very verbose

# token for Papertrail CLI access. used by realtime-updater
PAPERTRAIL_API_TOKEN=123456789012345678901234567890ab

# token for FullStory access. used to create links to FullStory
FULLSTORY_AUTH_TOKEN=123456789012345678901234567890ab

# databases/storage
ES_ADDRESS = <url to ElasticSearch database>
REDIS_ADDRESS = <url to Redis database>
USE_DOGPILE_CACHE = <True if we should use the dogpile cache>

# jira account to use for getting and creating issues
JIRA_SERVER = <url of the jira server. example: 'https://example.atlassian.net>
JIRA_BASIC_AUTH_USERNAME=name@server.com
JIRA_BASIC_AUTH_PASSWORD=hunter2
JIRA_PROJECT_KEY = <the project key of your JIRA project. example: 'SAN'>

# to download papertrail archives from s3
S3_BUCKET = <s3 bucket name. example: 'papertrail_log_bucket'>
S3_KEY_PREFIX = <prefix of papertrail logs in s3. example: 'papertrail/logs'>
AWS_ACCESS_KEY_ID=ASDFKJ87979898798798
AWS_SECRET_ACCESS_KEY=DFnkjdfkjdkFJDKFJkdsjfDFJKJKDFjkdfjkddkk # to download papertrail archives from s3
AWS_REGION=us-east-1 # to download papertrail archives from s3

# configuring the nginx AWS load balancer
AWS_SSL_CERT=arn:aws:acm:us-east-1:00000000000000:certificate/00000000-0000-0000-0000-000000000000
AWS_EXTRA_SECURITY_GROUP=sg-00000000000000
```
 - fill out .env.papertrail with config variables. requires these fields:
```
# logging
papertrail-destination=syslog://logs9.papertrailapp.com:12345
```

Then to start it all up:
 - `make fresh-deploy-to-k8s`

## Elasticsearch

For production, Elasticsearch database must be externally hosted. The IP of the
host server must be whitelisted.

Set up the mapping for the indexes by running `./scripts/setup-es-database.sh`

## Running tests
### Setup - install Python locally
 - pip install virtualenvwrapper
 - set up virtualenvwrapper
   - add this to your .bashrc:
```
export WORKON_HOME=$HOME/.virtualenvs
export PROJECT_HOME=$HOME/Devel
source /usr/local/bin/virtualenvwrapper.sh
```
 - `source .bashrc` to enable virtualenvwrapper
   - (if not installed) `yum install -y python34-virtualenv`
 - mkvirtualenv -p python3 <virtualenv name>
 - pip install -r requirements.txt

### Running
Once python is installed, tests can be run with
  - `make test`
  
### Setting up dockerhub
Use this command to use a username and password to log into docker. This allows 
you to push built images to  dockerhub:
- `docker login`

## S3 and Lambda
### Papertrail
Enable Papertrail's "archive to s3" feature
  - http://help.papertrailapp.com/kb/how-it-works/automatic-s3-archive-export/

### Lambda
## s3 Log File -> Application
Create an SNS topic to trigger a Lambda function whenever a Papertrail log file
is added to your s3 archive. The Lambda function should call our api
(/api/parse_s3) with the bucket and key of the log file.

Example lambda function:

```
import json
import requests
import urllib
import boto3

print('Loading function')

s3 = boto3.client('s3')

API_HOST_URL = # the URL of your app server
API_ENDPOINT = '/api/parse_s3'
API_REQUEST_URL = API_HOST_URL + API_ENDPOINT


def lambda_handler(event, context):
    # print("Received event: " + json.dumps(event, indent=2))

    # Get the object from the event and show its content type
    bucket = event['Records'][0]['s3']['bucket']['name']
    key = urllib.unquote_plus(event['Records'][0]['s3']['object']['key'].encode('utf8'))
    payload = {
        "bucket": bucket,
        "key": key,
    }
    print payload
    res = requests.post(API_REQUEST_URL, json=payload)
    if res.status_code in (200, 202):
        print('Successfully parsed "%s"' % key)
    else:
        print('Parse request received %s. data: "%s"' % (res.status_code, payload))
        raise Exception('parse request failed. payload: %s', payload)
```

## Jira ticket update -> Application
Configure a webhook in Jira to send a message every time a Jira ticket is
updated. Point that webhook at an API gateway. Point the API gateway at a lambda
function like this:
```
import json

import requests

import config


print('Loading function')

API_HOST_URL = config.API_HOST_URL
API_ENDPOINT = config.API_ENDPOINT
API_REQUEST_URL = API_HOST_URL + API_ENDPOINT


def lambda_handler(event, _):
    print("Received event: " + json.dumps(event))

    key = json.loads(event['body'])['issue']['key']
    payload = {
        'issue_key': key
    }
    print('making request: %s' % payload)
    res = requests.put(API_REQUEST_URL, json=payload)
    if res.status_code == 200:
        print('Successfully updated "%s"' % key)
    elif res.status_code == 202:
        print('Successfully queued "%s"' % key)
    else:
        print(
            'request failed. received response: %s. our payload: "%s"' %
            (res.status_code, payload)
        )
        raise
    
    return {'key': key}
```

Where config.py is
```
API_HOST_URL = # the URL of your app server
API_ENDPOINT = '/api/update_jira_db'
```

## Istio
The istio template was created with this command:
```
helm template istio-1.0.2/install/kubernetes/helm/istio --name istio --namespace istio-system --set tracing.enabled=true --set global.proxy.includeIPRanges="109.96.0.1/16" -x templates/sidecar-injector-configmap.yaml > kubernetes-istio/istio.yaml
```


CHANGELOG
Version 7.0.z
  - drop kubernetes support in favor of docker-compose

Version 6.2.z
  - add 'labels' to JiraIssues

Version 6.1.z
  - ignore certain tracebacks, more powerfully

Version 6.0.z
  - messages are posted to a more specific slack channel

Version 5.4.z
  - bump flask version

Version 5.3.z
  - add jaeger opentracing implementation

Version 5.2.z
  - add links to fullstory to output
  - add issue assignee name to Slack

Version 5.1.z
  - handle timestamps with missing 'T'
  - add socket_timeout to StrictRedis for the healthz endpoint
  - upgrade 'healthcheck' version

Version 5.0.z
  - add prometheus and istio monitoring
  - new 'requests' version

Version 4.1.z
  - include jira ticket with Slack postings
  - support Slack commenting on existing tickets
  - web comments use the same backend as Slack comments
  - redis 'seen it' set is replaced with many keys, so they can expire
  - add 'Add to Grader' option
  - add timing info back into logs
  - invalidate_cache is now a util

Version 4.0.z
  - save jira issue 'created and updated' datetime to ES
  - have a Filter for "tracebacks with tickets which haven't been updated recently"

Version 3.4.z
  - add ability to create a comment on a ticket, using a Slack search option
  - fixed lambda response, now clearing out Slack attachments using return values

Version 3.3.z
  - perform dogpile cache invalidation by deleting keys from redis
  - clear out Slack attachments using 'response_url'

Version 3.2.z
  - add mypy checking

Version 3.1.z
  - include a dropdown in Slack to select where to send the new ticket

Version 3.0.z
  - post new Tracebacks to slack
  - introduce simple class for 'TracebackPlusMetadata' object
  - version bump celery to 4.2.0
  - lots of other version bumps (requests, pylint, etc)

Version 2.8.0
  - revert changes from 2.6.0. we once again use the last line for matching

Version 2.7.0
  - don't provide a link to profile ids

Version 2.6.0
  - when making matches, we now ignore the final "AssertionError" line
    - this line can contain extra, superfluous information

Version 2.5.0
  - handle engine debug corner case

Version 2.4.0
  - correctly handle the case where we can't find the start of a traceback
  - handle tracebacks that are swallowed with the message 'Collected exception:'

Version 2.3.0
  - ignore log lines like this: `KeyError: 'Process worker 13'`

Version 2.2.0
  - long jira titles are no longer used

Version 2.1.0
  - new jira comment formatting
  - added a link to a "Jira formatted list" to index page

Version 2.0.0
  - add profile name parser

Version 1.4.0
  - add opentracing to elasticsearch calls

Version 1.3.0
  - add opentracing instrumentation lib. traces...
  -- top level requests
  -- redis requests

Version 1.2.0
  - jaeger tracing is now version 3.9.0 (finally built for python 3!)

Version 1.1.0
  - reduced logging output to avoid log recursion in papertrail
  - add http:// redirect

Version 1.0.0
  - Add Jaeger tracing


src/server.py
"""
    Run our web server.

    Provides endpoints for saving data to DB and for analyzing the data that's been saved.
"""
import datetime
import logging
import json
import os
import time
import traceback
import urllib

import certifi
import flask
import opentracing
import redis
from flask_bootstrap import Bootstrap
from flask_env import MetaFlaskEnv
from flask_kvsession import KVSessionExtension
from elasticsearch import Elasticsearch
from simplekv.memory.redisstore import RedisStore
from simplekv.decorator import PrefixDecorator

from opentracing_instrumentation.request_context import span_in_context

from common_util import (
    es_util,
    logging_util,
)
from lib.common import (
    cache_util,
)
from lib.jira import (
    jira_issue_aservice,
    jira_issue_db,
)

from lib.papertrail import (
    realtime_updater,
)
from lib.traceback import (
    traceback_db,
    traceback_formatter,
)
from webapp import (
    api_aservice,
    healthz,
    text_keys,
    tracing,
)
import tasks


# create app
ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
app = flask.Flask(__name__)

# get config from env vars
class EnvironmentVarConfig(metaclass=MetaFlaskEnv):
    ENV_LOAD_ALL = True # load all env variables

app.config.from_object(EnvironmentVarConfig)

logging_util.setup_logging()
logger = logging.getLogger()

# add bootstrap
Bootstrap(app)

# set up database
ES = Elasticsearch([app.config['ES_ADDRESS']], ca_certs=certifi.where())

# use redis for our session storage (ie: server side cookies)
REDIS = redis.StrictRedis(host=app.config['REDIS_ADDRESS'])
store = RedisStore(REDIS)
prefixed_store = PrefixDecorator('sessions_', store)
KVSessionExtension(prefixed_store, app)

# add route to /healthz healthchecks
healthz.add_healthcheck_endpoint(app, ES, REDIS)

# config
DEBUG_TIMING = True

FILTERS = ['All Tracebacks', 'Has Ticket', 'Has Open Ticket', 'No Ticket', 'No Recent Ticket']


@app.route("/", methods=['GET'])
def index():
    # parse the query params
    days_ago_raw = flask.request.args.get('days_ago')
    if days_ago_raw is not None:
        try:
            days_ago_int = int(days_ago_raw)
        except ValueError:
            return 'bad params', 400
    else:
        days_ago_int = 0
    filter_text = flask.request.args.get('filter')
    if filter_text is not None:
        filter_text = urllib.parse.unquote_plus(filter_text)
        if filter_text not in FILTERS:
            return 'bad filter: %s' % filter_text, 400
    if filter_text is None:
        filter_text = 'All Tracebacks'

    # create a set of traceback ids that match all the traceback texts the user has hidden
    span = flask.g.tracer_root_span
    tracer = opentracing.tracer
    with span_in_context(span):
        # TODO: do we need to set something to declare what this span block is doing?
        hidden_traceback_ids = set()
        if flask.session.get(text_keys.HIDDEN_TRACEBACK) is not None:
            for traceback_text in flask.session.get(text_keys.HIDDEN_TRACEBACK):
                for tb in traceback_db.get_matching_tracebacks(
                        ES, tracer, traceback_text, es_util.EXACT_MATCH, 10000
                ):
                    hidden_traceback_ids.add(tb.origin_papertrail_id)
            logger.info('found %s traceback ids we need to hide', len(hidden_traceback_ids))

    with span_in_context(span):
        span.set_tag('filter', filter_text)
        return api_aservice.render_main_page(ES, tracer, days_ago_int, filter_text, set())


@app.route("/api/parse_s3", methods=['POST'])
def parse_s3():
    """
        POST request to parse the data from a Papertrail log hosted on s3.

        Takes a JSON containing these fields:
        - bucket: the name of the s3 bucket containing the file. string
        - key: the filename of the file we should parse. must be in `bucket`. string

        All fields are required.

        Returns a 400 error on bad input. Returns a 202 after we queue the job to be run
        asyncronously.
    """
    # parse our input
    json_request = flask.request.get_json()
    if json_request is None or not all(k in json_request for k in ('bucket', 'key')):
        return 'missing params', 400
    bucket = json_request['bucket']
    key = json_request['key']

    logger.info("adding to s3 parse queue. bucket: '%s', key: '%s'", bucket, key)
    tasks.parse_log_file.delay(bucket, key)
    return 'job queued', 202


@app.route("/api/parse_s3_day", methods=['POST'])
def parse_s3_day():
    """
        POST request to parse the data from all Papertrail logs for one day

        Takes a JSON containing these fields:
        - date: day to parse. string, in YYYY-MM-DD form

        All fields are required.

        Returns a 400 error on bad input. Returns a 202 after we queue the jobs to be run
        asyncronously.
    """
    # parse our input
    json_request = flask.request.get_json()
    if json_request is None or not 'date' in json_request:
        return 'missing params', 400
    date_ = json_request['date']

    api_aservice.parse_s3_for_date(date_, app.config['S3_BUCKET'], app.config['S3_KEY_PREFIX'])
    return 'jobs queued', 202


@app.route("/api/parse_s3_date_range", methods=['POST'])
def parse_s3_date_range():
    """
        POST request to parse the data from all Papertrail logs for a date range

        Takes a JSON containing these fields:
        - start_date: starting day to parse. string, in YYYY-MM-DD form
        - end_date: starting day to parse. string, in YYYY-MM-DD form

        All fields are required.

        Returns a 400 error on bad input. Returns a 202 after we queue the jobs to be run
        asyncronously.
    """
    # parse our input
    json_request = flask.request.get_json()
    if (
            json_request is None
            or not 'start_date' in json_request
            or not 'end_date' in json_request
    ):
        return 'missing params', 400
    start_date_str = json_request['start_date']
    end_date_str   = json_request['end_date']

    # dates come in as a string - convert to datetime.date
    try:
        start_date = datetime.datetime.strptime(start_date_str, '%Y-%m-%d').date()
        end_date   = datetime.datetime.strptime(end_date_str,   '%Y-%m-%d').date()
    except ValueError:
        return 'failed to parse date from params', 400

    # iterate through all dates between start_date and end_date
    date_ = start_date
    while date_ <= end_date:
        api_aservice.parse_s3_for_date(date_, app.config['S3_BUCKET'], app.config['S3_KEY_PREFIX'])
        date_ += datetime.timedelta(days=1)

    return 'jobs queued', 202


@app.route("/realtime_update", methods=['POST'])
def realtime_update():
    """
        POST request to parse data directly from Papertrail in real time.

        Takes a JSON containing these fields:
        - end_time: datetime to parse. string, in '%Y-%m-%d %H:%M:%S' form

        All fields are optional.

        Returns a 400 error on bad input. Returns a 202 after we queue the job to be run
        asyncronously.
    """
    # parse our input. it's optional!
    end_time = None
    json_request = flask.request.get_json()
    if json_request is not None and 'end_time' in json_request:
        end_time = json_request['end_time']

    realtime_updater.enqueue(end_time)

    # this isn't exactly the best time to post to slack (since it won't include anything from this
    # realtime_update call), but it's as good a time as any. we expire this after a minute
    tasks.post_unticketed_tracebacks_to_slack.apply_async(tuple(), expires=60)
    return 'job queued', 202


@app.route("/hide_traceback", methods=['POST'])
def hide_traceback():
    json_request = flask.request.get_json()
    if json_request is None or 'traceback_text' not in json_request:
        logger.warning('invalid json detected: %s', json_request)
        return 'invalid json', 400
    traceback_text = json_request['traceback_text']
    our_set = flask.session.get(text_keys.HIDDEN_TRACEBACK)
    if our_set is None:
        our_set = set()
    # NOTE: changing the session implictly with session_object.add doesn't seem to work
    our_set.add(traceback_text)
    flask.session[text_keys.HIDDEN_TRACEBACK] = our_set
    return 'success'


@app.route("/restore_all", methods=['POST'])
def restore_all_tracebacks():
    flask.session[text_keys.HIDDEN_TRACEBACK] = None
    return 'success'


@app.route("/create_jira_ticket", methods=['POST'])
def create_jira_ticket():
    """
        Create a jira ticket with tracebacks that share the given traceback text

        Takes a json payload with these fields:
        - origin_papertrail_id: id of the traceback for which to create an issue

        The frontend is expecting this API to return a human readable string in the event of
        success (200 response code)
    """
    # get traceback id
    json_request = flask.request.get_json()
    if json_request is None or 'origin_papertrail_id' not in json_request:
        logger.warning('invalid json detected: %s', json_request)
        return 'invalid json', 400
    origin_papertrail_id = json_request['origin_papertrail_id']

    ticket_key = api_aservice.create_ticket(
        ES, origin_papertrail_id, None, reject_if_ticket_exists=False
    )

    # send toast message to user with the JIRA url
    url = jira_issue_aservice.get_link_to_issue(ticket_key)
    return 'Created ticket <a href="%s" class="alert-link">%s</a>' % (url, ticket_key)


@app.route("/jira_comment", methods=['POST'])
def jira_comment():
    """
        Save a comment in jira with the latest hits on this traceback

        Takes a json payload with these fields:
        - origin_papertrail_id: the id of the traceback to find papertrail matches for
        - issue_key: the jira issue key on which to leave the comment

        The frontend is expecting this API to return a human readable string in the event of
        success (200 response code)
    """
    # get the payload
    json_request = flask.request.get_json()
    if json_request is None:
        logger.warning('no json detected: %s', json_request)
        return 'missing json', 400
    if 'origin_papertrail_id' not in json_request:
        logger.warning('missing json field detected: %s', json_request)
        return 'missing origin_papertrail_id', 400
    if 'issue_key' not in json_request:
        logger.warning('missing json field detected: %s', json_request)
        return 'missing issue_key', 400
    origin_papertrail_id = json_request['origin_papertrail_id']
    issue_key = json_request['issue_key']

    api_aservice.create_comment_on_existing_ticket(ES, issue_key, origin_papertrail_id)
    url = jira_issue_aservice.get_link_to_issue(issue_key)
    return 'Created a comment on <a href="%s" class="alert-link">%s</a>' % (url, issue_key)


@app.route("/jira_formatted_list/<traceback_origin_id>", methods=['GET'])
def jira_formatted_list(traceback_origin_id):
    """
        Retrieves a formatted list fit for jira, with the latest hits on this traceback

        Takes a query param with these fields:
        - traceback_origin_id: the traceback to find papertrail matches for

        The frontend is expecting this API to return a human readable string in the event of
        success (200 response code)
    """
    try:
        traceback_id = int(traceback_origin_id)
    except ValueError:
        return 'bad traceback id', 400
    if not traceback_id:
        return 'missing traceback id', 400

    # get the referenced traceback
    tb = traceback_db.get_traceback(ES, traceback_id)

    # find a list of tracebacks that use the given traceback text
    tracebacks = traceback_db.get_matching_tracebacks(
        ES, opentracing.tracer, tb.traceback_text, es_util.EXACT_MATCH, 100
    )
    tracebacks.sort(key=lambda tb: int(tb.origin_papertrail_id), reverse=True)

    return (
        traceback_formatter.create_hits_list(tracebacks, traceback_formatter.jira_formatted_string),
        200,
        {'Content-Type': 'text/plain'}
    )


@app.route("/slack-callback", methods=['POST'])
def slack_callback():
    data = flask.request.get_data()
    if data is None:
        return 'empty data', 400

    parsed_data = urllib.parse.parse_qs(data)
    payload = json.loads(parsed_data[b'payload'][0])
    if 'actions' in payload:
        # it's a request to do an action
        action = payload['actions'][0]['name']
        logger.info('received action request for %s', action)
        if action == 'create_ticket':
            origin_papertrail_id = payload['callback_id']
            assign_to = payload['actions'][0]['selected_options'][0]['value']
            tasks.create_jira_ticket.delay(origin_papertrail_id, assign_to)

            # replace the slack message's CTA with a "done!" message
            original_message = payload['original_message']
            original_message['attachments'].pop() # destructive!
            original_message['attachments'].append(
                {
                    "text": "Creating ticket..."
                }
            )
            return flask.jsonify(original_message)
        elif action == 'add_to_existing_ticket':
            origin_papertrail_id = payload['callback_id']
            selected_ticket_key = payload['actions'][0]['selected_options'][0]['value']
            tasks.create_comment_on_existing_ticket.delay(selected_ticket_key, origin_papertrail_id)

            # replace the slack message's CTA with a "done!" message
            original_message = payload['original_message']
            original_message['attachments'].pop() # destructive!
            original_message['attachments'].append(
                {
                    "text": "Updating %s..." % selected_ticket_key
                }
            )
            return flask.jsonify(original_message)
        else:
            logger.error('unexpected slack callback action: %s', action)
            logger.warning('slack payload: %s', payload)
    elif 'name' in payload:
        # it's a request to get more info for a dropdown menu
        action = payload['name']
        logger.info('received dropdown request for %s', action)
        if action == 'add_to_existing_ticket':
            search_phrase = payload['value']
            options = {
                "options": list(jira_issue_aservice.search_matching_jira_tickets(ES, search_phrase))
            }
            return flask.Response(json.dumps(options), mimetype='application/json')
        else:
            logger.error('unexpected slack callback action: %s', action)
            logger.warning('slack payload: %s', payload)
    else:
        logger.error('unexpected slack callback json: %s', payload)
    return 'ok'


@app.route("/api/update_jira_db", methods=['PUT'])
def update_jira_db():
    """
        Update our database of jira issues.

        Takes a JSON payload with the following fields:
        - issue_key: if included, must be the key of the issue to update. if we see this key we
          only update the specified issue. example: SAN-1234
        - all: if included, must be the boolean value of True. if we see this key (and not
          issue_key) we update all jira issues.

        The JSON payload must have at least one field.
    """
    # parse our input
    json_request = flask.request.get_json()
    if json_request is None or not any(k in json_request for k in ('issue_key', 'all')):
        return 'invalid json', 400

    if 'issue_key' in json_request:
        # save the given issue to ES
        issue_key = json_request['issue_key']
        tasks.update_jira_issue.delay(issue_key, do_invalidate_cache=True)
        return 'job queued', 202
    else:
        if json_request['all'] != True:
            return 'invalid "all" json', 400
        # offload task onto our queue
        tasks.update_all_jira_issues.delay()
        return 'job queued', 202


@app.route("/api/invalidate_cache", methods=['PUT'])
@app.route("/api/invalidate_cache/<cache>", methods=['PUT'])
def invalidate_cache(cache=None):
    """
        Invalidate all (or a subset of) the dogpile function caches.
    """
    cache_util.invalidate_cache(cache)
    return 'success'


@app.route("/api/purge_celery_queue", methods=['PUT'])
def purge_celery_queue():
    num_tasks = REDIS.llen('celery')
    REDIS.delete('celery')
    logger.info('purged %s tasks', num_tasks)
    return 'success'


@app.route("/admin", methods=['GET'])
def admin():
    error = False
    num_jira_issues = None
    try:
        num_jira_issues = jira_issue_db.get_num_jira_issues(ES)
    except Exception:
        logger.warning('unable to find number of jira issues', exc_info=True)
        error = True
    num_celery_tasks = None
    try:
        num_celery_tasks = REDIS.llen('celery')
    except Exception:
        logger.warning('unable to find number of celery tasks', exc_info=True)
        error = True
    return flask.render_template(
        'admin.html',
        num_jira_issues=num_jira_issues,
        num_celery_tasks=num_celery_tasks,
        error=error,
    )


@app.before_first_request
def before_first_request():
    # add tracing
    tracing.initialize_tracer()


@app.before_request
def start_request():
    # log the request
    json_request = flask.request.get_json()
    json_str = '. json: %s' % str(json_request)[:100] if json_request is not None else ''
    logger.info(
        "received %s '%s' request from %s%s",
        flask.request.method,
        flask.request.full_path,
        flask.request.remote_addr,
        json_str,
    )

    # start an opentracing span
    headers = {}
    endpoint = flask.request.endpoint
    for k, v in flask.request.headers:
        headers[k.lower()] = v
    try:
        tracer = opentracing.tracer
        span_ctx = tracer.extract(opentracing.Format.HTTP_HEADERS, headers)
        span = tracer.start_span(operation_name=endpoint, child_of=span_ctx)
    except (opentracing.InvalidCarrierException, opentracing.SpanContextCorruptedException) as e:
        span = tracer.start_span(operation_name=endpoint, tags={"Extract failed": str(e)})
    span.set_tag('path', flask.request.full_path)
    span.set_tag('method', flask.request.method)
    span.set_tag('healthz', 'healthz' in flask.request.full_path)
    flask.g.tracer_root_span = span

    # record the start time so we can calculate timing info after the request is done
    flask.g.start_time = time.time()


@app.after_request
def after_request(response):
    """ End tracing and record a log after every request. """
    flask.g.tracer_root_span.finish()

    # figure out how long the request took
    try:
        time_diff = time.time() - flask.g.start_time
        time_diff_str = '%.2fs' % time_diff
    except AttributeError:
        logger.warning('unable to log request timing')
        time_diff_str = 'UNKNOWN'

    # This 'if' avoids the duplication of registry in the log, since that 500 is already logged via
    # @app.errorhandler.
    if response.status_code != 500:
        logger.info(
            "finished %s '%s' request from %s in %s. %s",
            flask.request.method,
            flask.request.full_path,
            flask.request.remote_addr,
            time_diff_str,
            response.status,
        )
    return response


@app.errorhandler(Exception)
def exceptions(_):
    """ Logging after every Exception. """
    tb = traceback.format_exc()
    logger.error(
        "error %s '%s' request from %s. 5xx internal server error\n%s",
        flask.request.method,
        flask.request.full_path,
        flask.request.remote_addr,
        tb,
    )

    return "Internal Server Error", 500


if __name__ == "__main__":
    app.run(debug=True, use_reloader=False, host='0.0.0.0', port=8000)
